### PR TITLE
Streamline `gardener-resource-manager` entrypoint and introduce component configuration

### PR DESCRIPTION
**Pull request by rfranzke**
**How to categorize this PR?**

/area dev-productivity
/kind cleanup technical-debt

**What this PR does / why we need it**:
This PR streamlines the `gardener-resource-manager` entrypoint as well as all controllers and webhooks so that it is aligned with other components like `garden{er-controller-manager,er-scheduler,let}`.

It also introduces a component config API for GRM (previously it was using CLI flags only).

**Which issue(s) this PR fixes**:
Part of #4251

**Release note**:

```breaking developer
The `gardener-resource-manager` component has been reworked entirely. It now uses a component config instead of CLI flags. Also, its Helm chart has been reworked entirely.
```

### CODE REVIEW

File: docs/concepts/resource-manager.md [hunk 1]
```diff
@@ -44,7 +54,7 @@ data:
     #   namespace: default
 ---
 apiVersion: resources.gardener.cloud/v1alpha1
-kind: ManagedResource
+kind: `ManagedResource`
```
Comment (oliver-goetz) [inline docs/concepts/resource-manager.md]: keeping the previous version looks better.
```suggestion
kind: ManagedResource
```

File: docs/concepts/resource-manager.md [hunk 2]
```diff
@@ -90,7 +100,7 @@ data:
 
 ---
 apiVersion: resources.gardener.cloud/v1alpha1
-kind: ManagedResource
+kind: `ManagedResource`
```
Comment (oliver-goetz) [inline docs/concepts/resource-manager.md]: same here

File: docs/concepts/resource-manager.md [hunk 3]
```diff
@@ -105,32 +115,37 @@ In this example the label `foo=bar` will be injected into the `Deployment` as we
...
-Condition checks will be skipped for such ManagedResources.
+If a `ManagedResource` is annotated with `resources.gardener.cloud/ignore=true` then it will be skipped entirely by the controller (no reconciliations or deletions of managed resources at all).
+However, when the `ManagedResource` itself is deleted (for example when a shoot is deleted) then the annotation is not respected and all resources will be deleted as usual.
+This feature can be helpful to temporarily patch/change resources managed as part of such `ManagedResource`.
+Condition checks will be skipped for such `ManagedResource`s.
 
 #### Modes
 
-The gardener-resource-manager can manage a resource in different modes. The supported modes are:
+The `gardener-resource-manager` can manage a resource in different modes. The supported modes are:
```
Comment (oliver-goetz) [inline docs/concepts/resource-manager.md]: This sentence says **different modes**, but there is only one (`Ignore`) listed below. It sounds as if other modes have been forgotten to be mentioned.

File: docs/concepts/resource-manager.md [hunk 4]
```diff
@@ -391,7 +407,7 @@ Any attempt to regenerate the token or creating a new such secret will again mak
 
 > You can opt-out of this behaviour for `ServiceAccount`s setting `.automountServiceAccountToken=false` by labeling them with `token-invalidator.resources.gardener.cloud/skip=true`.
 
-In order to enable the _TokenInvalidator_ you have to set `--token-invalidator-max-concurrent-workers` to a value larger than `0`.
+In order to enable the _TokenInvalidator_ you have to set `.controllers.tokenValidator.enabled=.webhooks.tokenValidator.enabled=true` in the component configuration.
```
Comment (oliver-goetz) [inline docs/concepts/resource-manager.md]:
```suggestion
In order to enable the _TokenInvalidator_ you have to set `.controllers.tokenInvalidator.enabled=true` and `.webhooks.tokenInvalidator.enabled=true` in the component configuration.
```

File: pkg/resourcemanager/apis/config/register.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+// limitations under the License.
+
+package config
+
+import (
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// GroupName is the group name use in this package.
```
Comment (oliver-goetz) [inline pkg/resourcemanager/apis/config/register.go]:
```suggestion
// GroupName is the group name used in this package.
```

File: pkg/resourcemanager/apis/config/validation/validation.go [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+	allErrs := field.ErrorList{}
+
+	allErrs = append(allErrs, validateSourceClientConnection(conf.SourceClientConnection, field.NewPath("sourceClientConnection"))...)
+	if conf.TargetClientConnection != nil {
+		allErrs = append(allErrs, validateTargetClientConnection(*conf.TargetClientConnection, field.NewPath("targetClientConnection"))...)
+	}
+	allErrs = append(allErrs, validateServerConfiguration(conf.Server, field.NewPath("server"))...)
+	allErrs = append(allErrs, componentbaseconfigvalidation.ValidateLeaderElectionConfiguration(&conf.LeaderElection, field.NewPath("leaderElection"))...)
+
+	if conf.LogLevel != "" && !sets.NewString(logger.AllLogLevels...).Has(conf.LogLevel) {
```
Comment (oliver-goetz) [inline pkg/resourcemanager/apis/config/validation/validation.go]: An empty LogLevel should not be allowed
```suggestion
	if !sets.NewString(logger.AllLogLevels...).Has(conf.LogLevel) {
```

File: pkg/resourcemanager/apis/config/validation/validation.go [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+		allErrs = append(allErrs, field.NotSupported(field.NewPath("logLevel"), conf.LogLevel, logger.AllLogLevels))
+	}
+	if conf.LogFormat != "" && !sets.NewString(logger.AllLogFormats...).Has(conf.LogFormat) {
```
Comment (oliver-goetz) [inline pkg/resourcemanager/apis/config/validation/validation.go]: Same for LogFormat
```suggestion
	if !sets.NewString(logger.AllLogLevels...).Has(conf.LogLevel) {
```

File: cmd/gardener-controller-manager/app/app.go [hunk 8]
```diff
@@ -123,8 +123,9 @@ func run(ctx context.Context, log logr.Logger, cfg *config.ControllerManagerConf
...
 	log.Info("Setting up manager")
 	mgr, err := manager.New(restConfig, manager.Options{
-		Logger:                  log,
-		Scheme:                  kubernetes.GardenScheme,
+		Logger: log,
+		Scheme: kubernetes.GardenScheme,
+
 		HealthProbeBindAddress:  fmt.Sprintf("%s:%d", cfg.Server.HealthProbes.BindAddress, cfg.Server.HealthProbes.Port),
 		MetricsBindAddress:      fmt.Sprintf("%s:%d", cfg.Server.Metrics.BindAddress, cfg.Server.Metrics.Port),
 		GracefulShutdownTimeout: pointer.Duration(5 * time.Second),
```
Comment (oliver-goetz) [inline cmd/gardener-controller-manager/app/app.go]:
```suggestion
		Logger: log,
		Scheme: kubernetes.GardenScheme,
		GracefulShutdownTimeout: pointer.Duration(5 * time.Second),

		HealthProbeBindAddress:  fmt.Sprintf("%s:%d", cfg.Server.HealthProbes.BindAddress, cfg.Server.HealthProbes.Port),
		MetricsBindAddress:      fmt.Sprintf("%s:%d", cfg.Server.Metrics.BindAddress, cfg.Server.Metrics.Port),
```
There is a tiny difference compare to the others 😅

**Review by oliver-goetz — COMMENTED** [review]
I'm currently in the middle of commit `Rework main entrypoint`. Some nits I discovered so far.

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+      namespace: {{ .Release.Namespace }}
+      path: /webhooks/invalidate-service-account-token-secret
+      port: 443
+  failurePolicy: Fail
+  matchPolicy: Exact
+  name: token-invalidator.resources.gardener.cloud
+  objectSelector:
+    matchLabels:
+      resources.gardener.cloud/purpose: token-invalidator
+  reinvocationPolicy: Never
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:28]: Deployment via botanist adds this namespace selector
```yaml
  namespaceSelector:
    matchExpressions:
    - key: gardener.cloud/purpose
      operator: NotIn
      values:
      - kube-system
      - kubernetes-dashboard
```

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+      operator: DoesNotExist
+    - key: app
+      operator: NotIn
+      values:
+      - gardener
+    - key: role
+      operator: NotIn
+      values:
+      - resource-manager
+  reinvocationPolicy: Never
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:71]: Deployment via botanist adds this namespace selector
```yaml
  namespaceSelector:
    matchExpressions:
    - key: gardener.cloud/purpose
      operator: NotIn
      values:
      - kube-system
      - kubernetes-dashboard
```

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 11]
```diff
@@ -0,0 +1,235 @@
...
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:113]: Deployment via botanist adds this namespace selector
```yaml
  namespaceSelector:
    matchExpressions:
    - key: gardener.cloud/purpose
      operator: NotIn
      values:
      - kube-system
      - kubernetes-dashboard
```

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 12]
```diff
@@ -0,0 +1,235 @@
...
+      operator: NotIn
+      values:
+      - gardener
+    - key: role
+      operator: NotIn
+      values:
+      - resource-manager
+    - key: topology-spread-constraints.resources.gardener.cloud/skip
+      operator: DoesNotExist
+  reinvocationPolicy: Never
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:155]: Deployment via botanist adds this namespace selector
```yaml
  namespaceSelector:
    matchExpressions:
    - key: gardener.cloud/purpose
      operator: NotIn
      values:
      - kube-system
      - kubernetes-dashboard
```

File: charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml [hunk 13]
```diff
@@ -0,0 +1,136 @@
+apiVersion: apps/v1
+kind: Deployment
+metadata:
+  name: gardener-resource-manager
+  namespace: {{ .Release.Namespace }}
+  labels:
+    app: gardener
+    role: resource-manager
+spec:
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml:9]: When deployed by botanist the labels look slightly different.
```yaml
  labels:
    app: gardener-resource-manager
```

However, the labels seem to be consistent in both scenarios.

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 14]
```diff
@@ -0,0 +1,235 @@
...
+    - key: projected-token-mount.resources.gardener.cloud/skip
+      operator: DoesNotExist
+    - key: app
+      operator: NotIn
+      values:
+      - gardener
+    - key: role
+      operator: NotIn
+      values:
+      - resource-manager
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:70]: Please see my comment at `gardener-resource-manager` deployment. 
The labels and selectors seem look consistently in the helm chart, but Botanist uses different labels.
```yaml
  labels:
    app: gardener-resource-manager
```

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 15]
```diff
@@ -0,0 +1,235 @@
...
+      path: /webhooks/seccomp-profile
+      port: 443
+  failurePolicy: Fail
+  matchPolicy: Exact
+  name: seccomp-profile.resources.gardener.cloud
+  objectSelector:
+    matchExpressions:
+    - key: seccompprofile.resources.gardener.cloud/skip
+      operator: DoesNotExist
+    - key: app
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:105]: Same here with the labels.

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml [hunk 16]
```diff
@@ -0,0 +1,235 @@
...
+      namespace: {{ .Release.Namespace }}
+      path: /webhooks/pod-topology-spread-constraints
+      port: 443
+  failurePolicy: Fail
+  matchPolicy: Exact
+  name: pod-topology-spread-constraints.resources.gardener.cloud
+  objectSelector:
+    matchExpressions:
+    - key: app
+      operator: NotIn
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:146]: Same here with the labels.

File: charts/gardener/resource-manager/charts/runtime/templates/configmap.yaml [hunk 17]
```diff
@@ -0,0 +1,149 @@
...
+    leaderElection:
+      leaderElect: {{ .Values.global.config.leaderElection.leaderElect }}
+      leaseDuration: {{ .Values.global.config.leaderElection.leaseDuration }}
+      renewDeadline: {{ .Values.global.config.leaderElection.renewDeadline }}
+      retryPeriod: {{ .Values.global.config.leaderElection.retryPeriod }}
+      resourceLock: {{ .Values.global.config.leaderElection.resourceLock }}
+      resourceName: {{ .Values.global.config.leaderElection.resourceName }}
+      resourceNamespace: {{ .Release.Namespace }}
+    logLevel: {{ .Values.global.config.leaderElection.logLevel }}
+    logFormat: {{ .Values.global.config.leaderElection.logFormat }}
```
Comment (oliver-goetz) [inline charts/gardener/resource-manager/charts/runtime/templates/configmap.yaml]: `logLevel` and `logFormat` under `leaderElection` looks strange. Maybe something like this.
```suggestion
    logLevel: {{ .Values.global.config.logLevel | default "info" }}
    logFormat: {{ .Values.global.config.logFormat | default "json" }}
```

**Review by oliver-goetz — CHANGES_REQUESTED** [review]
some bits and pieces in the helm chart. 
Otherwise 🚀

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 9 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:28]: Yes, but botanist is specific to seeds/shoots while this Helm chart is for general-purpose and does not make such assumptions.

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 10 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:71]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 11 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:113]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 12 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:155]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml (hunk 13 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/runtime/templates/deployment.yaml:9]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 14 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:70]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 15 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:105]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

File: charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml (hunk 16 above)
Comment (rfranzke) [reply charts/gardener/resource-manager/charts/application/templates/mutatingwebhookconfiguration.yaml:146]: ref https://github.com/gardener/gardener/pull/6865#discussion_r1005235810

**Review by oliver-goetz — APPROVED** [review]

**Pull request merged**